gh-125562: Update to documentation for exec* functions

jlownie opened the pull request:
# gh-125562: Update to documentation for exec* functions

This PR includes the following changes:

* Moving the documentation of the first argument to the `exec*` functions from the start of the Process Management section to the documentation of the `exec*` functions.
* Adding of headings within the `exec*` documentation.
* Moving the final paragraph of the `exec*` functions documentation to the "Location of the executable", which is where it is logically relevant.
* Adding links between the `system()`, `execl` and `spawnl` functions.
* Adding links to the `subprocess` module.

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--125565.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

jlownie committed:
gh-125562: Updates documentation for exec* functions

jlownie committed:
Minor change to section heading

jlownie committed:
gh-125562: Added links between the system(), spawnl() and execl() documentation sections

vadmium commented on Doc/library/os.rst:
```diff
@@ -4280,6 +4275,20 @@ to be ignored.
+   The various :func:`exec\* <execl>` functions take a list of arguments for the new
+   program loaded into the process.  In each case, the first of these arguments must
+   be the name of the program itself, and not an argument that a user may
+   have typed on a command line.
```
Since you already edited the second sentence, I’ll point out that often the first word on a command line is indeed the program name (e.g. shell command “echo bar”) so saying “not an argument . . . typed on a command line” is confusing if not wrong.

jlownie replied:
I'm open to any suggestions as to better wording?  That phrasing came from the original documentation, I agree it's not great but I found it difficult to come up with something better that I was confident was correct.

vadmium replied:
Here is my attempt:

The _exec_ functions take a list of arguments for the new program. The first of these is by convention the command name from the command line, rather than an argument to the command, and becomes _argv_[0] passed to the _main_ function of a C program. For example, _os.execv('/bin/echo', ['echo', 'foo', 'bar'])_ would only print _foo bar_; the _echo_ argument would seem to be ignored.

vadmium commented on Doc/library/os.rst:
```diff
@@ -4280,6 +4275,20 @@ to be ignored.
    :func:`sys.stdout.flush` or :func:`os.fsync` before calling an
    :func:`exec\* <execl>` function.
 
+   *Arguments*
```
```suggestion
   *Program arguments*
```

Differentiate from the other arguments to the _exec_ functions

vadmium commented on Doc/library/os.rst:
```diff
@@ -4289,29 +4298,35 @@ to be ignored.
    :func:`execlpe`, :func:`execvp`, and :func:`execvpe`) will use the
    :envvar:`PATH` environment variable to locate the program *file*.  When the
    environment is being replaced (using one of the :func:`exec\*e <execl>` variants,
    discussed in the next paragraph), the new environment is used as the source of
```
Now discussed three paragraphs down under _Environment variables_

vadmium replied:
```suggestion
   :envvar:`PATH` environment variable to locate the program *file*.  When the
   environment is being replaced (see *Environment variables*
   below), the new environment is used as the source of
```

vadmium commented on Doc/library/os.rst:
```diff
@@ -4280,6 +4275,20 @@ to be ignored.
+   be the name of the program itself, and not an argument that a user may
+   have typed on a command line.
+
+   For the C programmer, this is the ``argv[0]``
```
I’m not sure breaking the paragraph works. “This” refers to the name of the program (topic of the last sentence of the last paragraph), not the overall Program Arguments heading.

jlownie replied:
I decided to retain the paragraph break and instead reworded the first sentence in the paragraph.  I'm not entirely sure what the documentation is getting at with the second sentence in the paragraph.  I'm open to further suggestions for improving this paragraph - I feel it could be improved but I'm not sure how.

vadmium commented on Doc/library/os.rst:
```diff
@@ -4289,29 +4298,35 @@ to be ignored.
+   the :envvar:`PATH` variable.
+
+   The other variants, :func:`execl`, :func:`execle`,
    :func:`execv`, and :func:`execve`, will not use the :envvar:`PATH` variable to
```
```suggestion
   The variants without "p" (:func:`execl`, :func:`execle`,
   :func:`execv`, and :func:`execve`) will not use the :envvar:`PATH` variable to
```

vadmium commented on Doc/library/os.rst:
```diff
@@ -4323,6 +4338,11 @@ to be ignored.
+   .. seealso::
+      The :mod:`subprocess` module.
+
+      The :func:`system` and :func:`spawnl` functions also execute a system command.
```
```suggestion
      The :func:`system` and :func:`spawn <spawnl>` functions also execute a new program.
```

vadmium commented on Doc/library/os.rst:
```diff
@@ -4269,6 +4260,10 @@ to be ignored.
 
+   (Note that the :mod:`subprocess` module provides more powerful facilities for
+   spawning new processes and retrieving their results; using that module is
+   preferable to using these functions.)
```
_Subprocess_ isn’t a replacement for _exec_; it does not replace the existing process, as far as I know.

jlownie replied:
Text removed.

vadmium commented on Doc/library/os.rst:
```diff
@@ -4897,6 +4918,10 @@ written in Python, such as a mail server's external command delivery program.
+   .. seealso::
+      The :mod:`subprocess` module.
+
+      The :func:`system` and :func:`execl` functions also execute a system command.
```
```suggestion
      The :func:`system` and :func:`exec <execl>` functions also execute a new program.
```

vadmium commented on Doc/library/os.rst:
```diff
@@ -5013,6 +5037,10 @@ written in Python, such as a mail server's external command delivery program.
+   .. seealso::
+      The :mod:`subprocess` module.
+
+      The :func:`execl` and :func:`spawnl` functions also execute a system command.
```
```suggestion
      The :func:`exec <execl>` and :func:`spawn <spawnl>` functions also execute a system command.
```

vadmium reviewed (commented):
Putting the bracketed deprecation paragraphs at the top of each description is not great. Maybe move them down, if needed at all.

jlownie committed:
Update Doc/library/os.rst

Co-authored-by: Martin Panter <vadmium@users.noreply.github.com>

jlownie committed:
gh-125562: Removed incorrect text regarding using subprocess.run instead of execl

jlownie committed:
gh-125562: Altered wording before 'argv[0]' in execl documentations

jlownie committed:
Apply suggestions from code review

Co-authored-by: Martin Panter <vadmium@users.noreply.github.com>

jlownie committed:
Merge branch 'exec_doco' of https://github.com/jlownie/cpython into exec_doco

jlownie committed:
Merge branch 'main' into exec_doco

vadmium commented on Doc/library/os.rst:
```diff
@@ -4280,6 +4271,20 @@ to be ignored.
+   ['foo', 'bar'])`` will only print ``bar`` on standard output; ``foo`` will seem
+   to be ignored.
+
+   *Fixed and variable numbers of arguments*
```
Maybe better off without this heading? The paragraph still falls under “Program arguments”, for a total of only two or three paragraphs.

jlownie replied:
Agreed, done.

jlownie committed:
Merge branch 'main' into exec_doco

jlownie committed:
pythongh-125562: Rewording of documentation for execl function

Incorporated vadmium's suggestion to reference the environment variables section

Co-authored-by: Martin Panter <vadmium@users.noreply.github.com>

jlownie committed:
pythongh-125562: Rewording of documentation for execl function

jlownie committed:
Merge branch 'exec_doco' of https://github.com/jlownie/cpython into exec_doco

jlownie committed:
Merge branch 'main' into exec_doco

jlownie committed:
Merge branch 'main' into exec_doco

jlownie committed:
Removed trailing space

jlownie committed:
Removed trailing whitespace
